c2rust-analyze/src/context.rs

```diff
@@ -98,13 +114,44 @@ pub struct AnalysisCtxt<'a, 'tcx> {
     /// those `PointerId`s consistent, the `Rvalue`'s type must be stored rather than recomputed on
     /// the fly.
     pub rvalue_tys: HashMap<Location, LTy<'tcx>>,
+
+    /// [`Location`]s of const ref [`rvalue_tys`](Self::rvalue_tys).
+    const_ref_locs: Vec<Location>,
+
     next_ptr_id: NextLocalPointerId,
 }
 
+impl<'a, 'tcx> AnalysisCtxt<'_, 'tcx> {
+    pub fn const_ref_tys(&'a self) -> impl Iterator<Item = LTy<'tcx>> + 'a {
+        self.const_ref_locs.iter().map(|loc| self.rvalue_tys[loc])
+    }
+
+    pub fn const_ref_perms(
+        &'a self,
+    ) -> impl Iterator<Item = (PointerId, PermissionSet)> + PhantomLifetime<'tcx> + 'a {
```

Comment:
Is the `PhantomLifetime<'tcx>` actually necessary here?  I don't think the returned iterator borrows anything with lifetime `'tcx`.  `self.const_ref_tys()` returns an iterator that doesn't borrow `'tcx`, and the closure provided to `map` doesn't capture anything.

Reply:
It is.  `self.const_ref_tys()` returns `LTy<'tcx>`s, and so I need that `'tcx` in the return type somewhere.  Since there's no more `LTy`, I need to stick it somewhere.